webui/react/src/hooks/useSettings.tsx

```diff
@@ -0,0 +1,332 @@
+import { Type } from 'io-ts';
+import queryString from 'query-string';
+import React, { createContext, useCallback, useContext, useEffect, useMemo, useState } from 'react';
+import { useNavigate } from 'react-router-dom-v5-compat';
+
+import { useStore } from 'contexts/Store';
+import { getUserSetting, updateUserSetting } from 'services/api';
+import { V1UserWebSetting } from 'services/api-ts-sdk';
+import { UpdateUserSettingParams } from 'services/types';
+import { Primitive } from 'shared/types';
+import { isEqual } from 'shared/utils/data';
+import { ErrorType } from 'shared/utils/error';
+import handleError from 'utils/error';
+
+/*
+ * UserSettingsState contains all the settings for a user
+ * across the application. Each key identifies a unique part
+ * of the interface to store settings for.
+ */
+type UserSettingsState = Map<string, Settings>;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Settings = { [key: string]: any }; //TODO: find a way to use a better type here
+
+export enum BaseType {
+  Boolean = 'Boolean',
+  Float = 'Float',
+  Integer = 'Integer',
+  Object = 'Object',
+  String = 'String',
+}
+
+export interface SettingsConfigProp<A> {
+  defaultValue: A;
+  skipUrlEncoding?: boolean;
+  storageKey: string;
+  type: {
+    baseType: Type<A>;
+    isArray?: boolean;
+  };
+}
+
+export interface SettingsConfig<T> {
+  applicableRoutespace: string;
+  settings: { [K in keyof T]: SettingsConfigProp<T[K]> };
+  storagePath: string;
+}
+
+type UserSettingsContext = {
+  state: UserSettingsState;
+  update: (key: string, value: Settings) => void
+}
+
+interface UserSettingUpdate extends UpdateUserSettingParams {
+  userId: number;
+}
+
+export type UpdateSettings = (updates: Settings, shouldPush?: boolean) => Promise<void>;
+export type ResetSettings = (settings?: string[]) => void;
+
+export type UseSettingsReturn = {
+  activeSettings: (keys?: string[]) => string[];
+  resetSettings: ResetSettings;
+  settings: Settings;
+  updateSettings: UpdateSettings;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const UserSettings = createContext<UserSettingsContext>({ state: new Map(), update: () => { } });
+
+const SettingsProvider: React.FC = ({ children }) => {
+  const [canceler] = useState(new AbortController());
+  const [apiSettings, setApiSettings] = useState<V1UserWebSetting[]>([]);
+
+  useEffect(() => {
+    getUserSetting({}, { signal: canceler.signal })
+    .then((response) => {
+      setApiSettings(response.settings);
+    });
+
+    return () => canceler.abort();
+  }, [canceler]);
+
+  const settingsState = useMemo(() => {
```

Comment:
The way you're using this is a little confusing. I think you're basically ignoring the state subscriptions except for the initial fetching. This makes sense to me since you only want to trigger re-renders for individual items in the map. In that case I would get rid of `apiSettings` and just put this map in a `useRef` that you then populate with the initial settings.

I think we maybe should have an `isLoading` state boolean too which will trigger the one global re-render we need. 

So that would be like:
```typescript
const settingsState = useRef<Map<string, any>(new Map());
const [isLoading, setIsLoading] = useState(true);

...

<UserSettings.Provider value={{
      state: settingsState,
      update,
      isLoading,
    }}>...
```

Then we can also have proper loading states in the components that use `useSettings`.